rfcs/testing.md

```diff
@@ -0,0 +1,75 @@
+# RFC: Testing standards
+
+---
+
+_@andrefcdias_
+
+## Summary
+
+The purpose of this RFC is to standardize our testing process and clarify the purpose of each of the testing layers we have. It is supposed to define what is the intended result from tests and why they are required.
+
+## Background
+
+This RFC came from the discussion of unit testing styles in components such as Flex and Text and the debate of having a valuable test vs a "correct" one.
+
+## Problem statement
+
+At the moment, everyone has their own testing standards and there is no common approach to problems like components that only apply styles and need to be tested.
+There is no majority decision on the solution for this problem, being split between Visual Regression tests (VR tests) and Unit Tests (UT).
+
+By defining a common approach, we can make the testing process less ambiguous, as well as provide a guide for the community to follow.
+
+## Detailed Design or Proposal
+
+To start, this is the proposal for defining the goals of each testing layer.
+
+- **Conformance**
+
+  Ensuring the consistency of our components. They test that props like `as`, `ref`, `className` are supported and that the component is rendered.
+
+  These can be achieved simply through the `isConformant` function generated when creating a converged component.
+
+- **Snapshot**
+
+  Stricly DOM structure consistency enforcing. This will guarantee we do not introduce breaking changes for users who expect a certain structure to be rendered by our components.
+
+  They are achieved with Jest's `expect(component).toMatchInlineSnapshot(...)`. This is preferred over `.toMatchSnapshot` as it is immediately explicit what is expected from the test.
+
+- **Unit**
+
+  Focus on testing the behaviour or functionality. We should use them to guarantee the outcome, wheter this is a styling, behaviour or functionality change.
+  We should try to focus on behaviour, when possible, but mainly on the expected results for the code.
+
+  These are all the additional unit tests defined in your `*.test.ts[x]` file.
+
+- **Integration**
```

Comment:
Having both unit+integration tests seems to be very strange for me, and often even in applications the lines are blurred. I would say that we should just have a single unit test category

Reply:
I agree, perhaps two categories is misleading but with this RFC I had in mind to cover all kinds of tests and standardize how we do them.
Do you think some rephrasing would help? What I want to say is exactly what you mentioned, our Unit and Integration tests go hand in hand.

Reply:
Well especially since we currently **don't** have any such thing as integration tests, we should do away with the whole thing until we know what we want from such a thing.

I would rather say that we should mock as little as possible in our unit tests, since we are a library, our dependencies really should be tested as much as possible.

Reply:
Ah I understand now. By this I meant integration testing internally our dependencies, i.e. `make-styles` without mocking it, and not integration of everything as a whole somewhere.
Would it make sense to remove altogether or merge it in Unit Tests?

Reply:
My proposal: remove integration altogether and just add that in our unit tests we aim to mock as a little as possible